### ISSUE TITLE
scli is not sending messages

### ISSUE THREAD

**Issue by bluecub** [isamert/scli#86]
Hi, 
I installed scli on freeBSD using package. It installed signal-cli.
I linked the device, and was able to send messages from signal-cli successfully.
I stated the scli and I can see all my contacts but the messages I send won't be delivered. I ran the app in the debug mode and I found this error in .local/share/scli/log file.

```
ERROR:root:stderr: org.freedesktop.dbus.exceptions.DBusException: Failed to connect to bus: No such file or directory
ERROR:root:stderr: at org.freedesktop.dbus.connections.AbstractConnection.<init>(AbstractConnection.java:166)
	at org.freedesktop.dbus.connections.impl.DBusConnection.<init>(DBusConnection.java:334)
	at org.freedesktop.dbus.connections.impl.DBusConnection.getConnection(DBusConnection.java:149)
	at org.freedesktop.dbus.connections.impl.DBusConnection.getConnection(DBusConnection.java:169)
	at org.freedesktop.dbus.connections.impl.DBusConnection.getConnection(DBusConnection.java:228)
	at org.freedesktop.dbus.connections.impl.DBusConnection.getConnection(DBusConnection.java:184)
	at org.asamk.signal.commands.DaemonCommand.handleCommand(DaemonCommand.java:51)
	at org.asamk.signal.Main.handleCommands(Main.java:181)
	at org.asamk.signal.Main.handleCommands(Main.java:132)
	at org.asamk.signal.Main.main(Main.java:64)
Caused by: java.io.IOException: No such file or directory
	at jnr.unixsocket.UnixSocketChannel.doConnect(UnixSocketChannel.java:130)
	at jnr.unixsocket.UnixSocketChannel.connect(UnixSocketChannel.java:139)
	at jnr.unixsocket.UnixSocketChannel.open(UnixSocketChannel.java:68)
	at org.freedesktop.dbus.connections.transports.UnixSocketTransport.connect(UnixSocketTransport.java:59)
	at org.freedesktop.dbus.connections.transports.TransportFactory.createTransport(TransportFactory.java:44)
ERROR:root:stderr: at org.freedesktop.dbus.connections.AbstractConnection.<init>(AbstractConnection.java:161)
	... 9 more
INFO:root:send_message:{'source': '+614xxxxxxxxx', 'target': '+614xxxxxxxxx', 'timestamp': 1599564548760, 'dataMessage': {'message': 'Hi', 'attachments': [], 'timestamp': 1599564548760}}
ERROR:root:proc: cmd:`dbus-send --session --type=method_call --print-reply --dest=org.asamk.Signal /org/asamk/Signal org.asamk.Signal.sendMessage string:Hi array:string: string:+61413761863 1>&2; echo $$ $?`; return_code:1; output:"Error org.freedesktop.DBus.Error.Spawn.ExecFailed: Failed to execute program org.asamk.Signal: No such file or directory"

```

Any idea how to solve this?

**Comment by exquo** [comment]
Looks like it does not see signal-cli's dbus service. Do you get anything in reply if you execute

	dbus-send --print-reply --dest=org.freedesktop.DBus  /org/freedesktop/DBus org.freedesktop.DBus.ListNames  | grep -i signal

or the same command with a `--system` flag (`dbus-send --system ...`)?

There are three ways to send a message through signal-cli (that I know of):
- `signal-cli -u +USERNAME send -m 'MESSAGE HERE' +RECIPIENT`

Or, run signal-cli in the daemon mode:

	signal-cli -u +USERNAME daemon

and then
- `signal-cli --dbus send -m 'MESSAGE HERE' +RECIPIENT`
or
- `dbus-send --session --type=method_call --print-reply --dest="org.asamk.Signal" /org/asamk/Signal org.asamk.Signal.sendMessage string:'MESSAGE HERE' array:string: string:+RECIPIENT`
(from https://github.com/AsamK/signal-cli/wiki/DBus-service)

I assume you've tried the first one. Could you try the other two and post back on how it goes?

Possibly related, although it's marked as resolved: 
AsamK/signal-cli#287

**Comment by bluecub** [comment]
Hi there,
This one signal-cli -u +USERNAME send -m 'MESSAGE HERE' +RECIPIENT woked like yesterday

signal-cli -u +USERNAME daemon failed
```
bsd:~/bin % signal-cli -u +Username daemon
org.freedesktop.dbus.exceptions.DBusException: Failed to connect to bus: No such file or directory
	at org.freedesktop.dbus.connections.AbstractConnection.<init>(AbstractConnection.java:166)
	at org.freedesktop.dbus.connections.impl.DBusConnection.<init>(DBusConnection.java:334)
	at org.freedesktop.dbus.connections.impl.DBusConnection.getConnection(DBusConnection.java:149)
	at org.freedesktop.dbus.connections.impl.DBusConnection.getConnection(DBusConnection.java:169)
	at org.freedesktop.dbus.connections.impl.DBusConnection.getConnection(DBusConnection.java:228)
	at org.freedesktop.dbus.connections.impl.DBusConnection.getConnection(DBusConnection.java:184)
	at org.asamk.signal.commands.DaemonCommand.handleCommand(DaemonCommand.java:51)
	at org.asamk.signal.Main.handleCommands(Main.java:181)
	at org.asamk.signal.Main.handleCommands(Main.java:132)
	at org.asamk.signal.Main.main(Main.java:64)
Caused by: java.io.IOException: No such file or directory
	at jnr.unixsocket.UnixSocketChannel.doConnect(UnixSocketChannel.java:130)
	at jnr.unixsocket.UnixSocketChannel.connect(UnixSocketChannel.java:139)
	at jnr.unixsocket.UnixSocketChannel.open(UnixSocketChannel.java:68)
	at org.freedesktop.dbus.connections.transports.UnixSocketTransport.connect(UnixSocketTransport.java:59)
	at org.freedesktop.dbus.connections.transports.TransportFactory.createTransport(TransportFactory.java:44)
	at org.freedesktop.dbus.connections.AbstractConnection.<init>(AbstractConnection.java:161)
	... 9 more
```


bsd:~/bin % dbus-send --session --type=method_call --print-reply --dest="org.asamk.Signal" /org/asamk/Signal org.asamk.Signal.sendMessage string:'MESSAGE HERE' array:string: string:+RECIPIENT
Error org.freedesktop.DBus.Error.Spawn.ExecFailed: Failed to execute program org.asamk.Signal: No such file or directory




I already did the below when I installed it (as per the package message)
```
signal_cli service (using D-Bus system bus)
===========================================

There is an rc(8) script available for signal-cli if it's desired to use
signal-cli in a daemon mode with the D-Bus system bus.

First, configure the service in rc.conf(5):

    sysrc signal_cli_enable="YES"
    sysrc signal_cli_phone_number="+USERNAME"

Before starting the service, it is needed to either register a new Signal
account or to link this machine with an existing account. It is possible to
create a linking URL this way:

    service signal_cli runcli link -n DEVICENAME
```

**Comment by exquo** [comment]
Looks to me like something's wrong with dbus, probably not even signal-cli specific. Do you have any ouput from

	dbus-send --print-reply --dest=org.freedesktop.DBus  /org/freedesktop/DBus org.freedesktop.DBus.ListNames

**Comment by bluecub** [comment]
bsd:~ % dbus-send --print-reply --dest=org.freedesktop.DBus  /org/freedesktop/DBus org.freedesktop.DBus.ListNames
method return time=1599739202.759277 sender=org.freedesktop.DBus -> destination=:1.34 serial=3 reply_serial=2
   array [
      string "org.freedesktop.DBus"
      string ":1.7"
      string ":1.8"
      string ":1.9"
      string "org.a11y.Bus"
      string ":1.10"
      string ":1.34"
      string ":1.1"
      string ":1.2"
      string ":1.3"
      string ":1.4"
      string ":1.5"
      string ":1.6"
   ]

**Comment by exquo** [comment]
That list looks a bit barren. The output on my fairly minimal debian includes

	org.freedesktop.Notifications
	org.freedesktop.systemd1
	org.gtk.vfs.Daemon
	org.pulseaudio.Server

and many others. Maybe dbus is only half-functioning on your system? Reporting to be running, but not registering new services?

You could try to register a simple dbus service and see if that works. Maybe something along the lines described here 
https://stackoverflow.com/questions/34482691/register-a-hello-world-dbus-service-object-and-method-using-python
Or you could test another program that communicates over dbus and is known to run on BSD. Maybe it's just a matter of putting `dbus_enable=YES` in `/etc/rc.conf` and restarting. But I know next to nothing about dbus and even less about BSD.

In any case, the problem is "below" `scli`'s purview. You could open an issue against signal-cli's BSD version, or email the developer on freshports and link to this discussion. But I would guess it's not about signal-cli either, just about the dbus setup. I'll close this issue, but feel free to add any new comments below.

**Comment by exquo** [comment]
There has just now been a new version of signal-cli released (v0.6.9) that says in release notes
> dbus functionality now works on FreeBSD

I guess that means that it's supposed to work without the changes from the bsd 'port' version. See if that helps.

**Comment by bluecub** [comment]
Thanks so much @exquo I really appreciate the time you put to try to help solve the issue.
DBus is already running my I don't have DE (just i3) so definitely no gtk and FreeBSD doesn't come with systemD and org.pulseaudio.Server.

I will upgrade the cli and hopefully it works. Thanks so much 👍

**Comment by exquo** [comment]
Sure thing!

**Comment by antranigv** [comment]
Hi there! I had a similar issue which I ended up solving by setting the environment variable, I blogged about it [here](https://antranigv.am/weblog_en/posts/freebsd-signal-cli-scli/).

TL;DR: Find the `DBUS_SESSION_BUS_ADDRESS` and set it in your shell, then execute `scli` :-) 

P.S. Thanks for creating scli!

**Comment by bluecub** [comment]
@antranigv  cheers mate, I will give it ago and let you know how it goes.

**Comment by exquo** [comment]
Thanks for the tip, @antranigv!

I have added it to signal-cli's wiki [here](https://github.com/AsamK/signal-cli/wiki/Quickstart#user-content-resolving-dbus-issues-on-freebsd) and opened an issue about this in AsamK/signal-cli#363

**Comment by bluecub** [comment]
@exquo  and @antranigv  That solved the issue for me..... Thanks for the tip and thanks @exquo for the great work you;'ve put into scli
